cmd/tailscale/cli/serve.go

```diff
@@ -101,7 +134,66 @@ func (e *serveEnv) stdout() io.Writer {
 }
 
 func (e *serveEnv) runServe(ctx context.Context, args []string) error {
-	panic("TODO")
+	if len(args) != 3 {
+		return flag.ErrHelp
+	}
+
+	path := args[0]
+
+	h := new(ipn.HTTPHandler)
+
+	switch args[1] {
+	case "path":
+		h.Path = args[2]
+	case "proxy":
+		portStr := args[2]
+		_, err := strconv.ParseUint(portStr, 10, 16)
+		if err != nil {
+			printf("error: invalid port %q\n\n", portStr)
+			return flag.ErrHelp
+		}
+		// TODO(shayne): handle protocols
+		h.Proxy = "localhost:" + args[2]
+	case "text":
+		h.Text = args[2]
+	default:
+		// TODO(shayne): fix, tell why
+		return flag.ErrHelp
+	}
+
+	sc, err := e.getServeConfig(ctx)
+	if err != nil {
+		// TODO(shayne): logging?
+		return err
+	}
+
+	dnsName, err := e.getSelfDNSName(ctx)
+	if err != nil {
+		return err
+	}
+	dnsName = strings.TrimSuffix(dnsName, ".")
+
+	var key ipn.HostPort = ipn.HostPort(net.JoinHostPort(dnsName, "443"))
+
+	if sc != nil && sc.Web != nil && sc.Web[key] != nil &&
+		sc.Web[key].Handlers[path] != nil {
+		if (h.Proxy != "" && sc.Web[key].Handlers[path].Proxy == h.Proxy) ||
+			(h.Text != "" && sc.Web[key].Handlers[path].Text == h.Text) ||
+			(h.Path != "" && sc.Web[key].Handlers[path].Path == h.Path) {
+			// Nothing to do.
+			return nil
+		}
+	}
```

Comment:
might be easier to blindly make the change to another copy of sc and then DeepEqual them to see if you need to call setServeConfig